src/transcript.rs

```diff
@@ -43,3 +44,86 @@ impl<F: Field> Hasher<F> for DummyHash<F> {
         tmp
     }
 }
+
+/// A transcript that can absorb points from both the base field and scalar
+/// field of a curve
+#[derive(Debug, Clone)]
+pub struct Transcript<C: CurveAffine, HBase, HScalar>
+where
+    HBase: Hasher<C::Base>,
+    HScalar: Hasher<C::Scalar>,
+{
+    // Hasher over the base field
+    base_hasher: HBase,
+    // Hasher over the scalar field
+    scalar_hasher: HScalar,
+    // Indicates if scalar(s) has been hashed but not squeezed
+    scalar_needs_squeezing: bool,
+    // PhantomData
+    _marker: PhantomData<C>,
+}
+
+impl<C: CurveAffine, HBase: Hasher<C::Base>, HScalar: Hasher<C::Scalar>>
+    Transcript<C, HBase, HScalar>
+{
+    /// Initialise a new transcript with Field::one() as keys
+    /// in both the base_hasher and scalar_hasher
+    pub fn new() -> Self {
+        let base_hasher = HBase::init(C::Base::one());
+        let scalar_hasher = HScalar::init(C::Scalar::one());
+        Transcript {
+            base_hasher,
+            scalar_hasher,
+            scalar_needs_squeezing: false,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Initialise a new transcript with some given base_hasher and
+    /// scalar_hasher
+    #[cfg(test)]
+    pub(crate) fn init_with_hashers(base_hasher: &HBase, scalar_hasher: &HScalar) -> Self {
+        Transcript {
+            base_hasher: base_hasher.clone(),
+            scalar_hasher: scalar_hasher.clone(),
+            scalar_needs_squeezing: false,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Absorb a curve point into the transcript by absorbing
+    /// its x and y coordinates
+    pub fn absorb_point(&mut self, point: &C) -> Result<(), ()> {
+        let tmp = point.get_xy();
+        if bool::from(tmp.is_none()) {
+            return Err(());
+        };
+        let tmp = tmp.unwrap();
+        self.base_hasher.absorb(tmp.0);
+        self.base_hasher.absorb(tmp.1);
+        Ok(())
+    }
+
+    /// Absorb a base into the base_hasher
+    pub fn absorb_base(&mut self, base: C::Base) {
+        self.base_hasher.absorb(base);
```

Comment:
Is there a possibility that absorbing two bases in a row could be confused with absorbing a point (in terms the transcript malleability)? Where do we use `absorb_base`, and do we need to differentiate between absorbing individual bases and points (with e.g. some kind of domain separator)?

cc @ebfull @daira

Reply:
Non-blocking; open an issue if we don't answer this here.

Reply:
Opened https://github.com/zcash/halo2/issues/69.